README.md
# Overview
[![Build Status](https://travis-ci.org/ww-tech/primrose.svg?branch=master)](https://travis-ci.org/ww-tech/primrose)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/primrose.svg)](https://pypi.python.org/pypi/primrose/)
[![PyPI version](https://badge.fury.io/py/primrose.svg)](https://badge.fury.io/py/primrose)
[![PyPI license](https://img.shields.io/pypi/l/primrose.svg)](https://pypi.python.org/pypi/primrose/)
[![Docs status](https://img.shields.io/website/https/ww-tech.github.io/primrose?down_color=red&down_message=docs&label=docs&up_color=success&up_message=up)](https://ww-tech.github.io/primrose/)


<p align="center">
   <img src="img/primrose_logo.png" width="100">
</p>

## Primrose at a glance

`Primrose` is a simple **Python** framework for executing **in-memory** workflows defined by directed acyclic graphs (**DAGs**) via configuration files. Data in `primrose` flows from one node to another while **avoiding serialization**, except for when explicitly specified by the user. `Primrose` nodes are designed for **simple batch-based machine learning workflows**, which have datasets small enough to fit into a single machine's memory.

## Table of Contents
We suggest reading the documentation in the following order:

 - Overview and motivation for `primrose`&mdash;this file.
 - [Getting Started](README_GETTING_STARTED.md): run your first `primrose` jobs.
 - [DAG Configurations](README_DAG_CONFIG.md): `primrose` adopts a configuration-as-code paradigm. This section introduces `primrose` configuration files.
 - [Metadata](README_METADATA.md): this covers more advanced options of the configuration files.
 - [Command Line Interface (CLI)](README_CLI.md): run commands using the CLI.
 - [Developer Notes](README_DEVELOPER_NOTES.md): how to create your own new Node classes.
 - [DataObject](README_DATAOBJECT.md): a deep dive into `DataObject`, the core data handling and book-keeping object.

## Introduction

 `Primrose` is a Python framework for quick, simple machine learning and recommender deployments developed by the data science team at [WW](https://www.weightwatchers.com/us/). It is essentially a workflow management tool which is specialized for the needs of machine learning tasks with small to medium sized datasets (&le; 100GB). Like many orchestration tools, `Primrose` *nodes* are defined in a [directed-acyclic-graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) which defines dependencies and control flow.

 Here's an example DAG showing data cleaning, model training, and model serialization:

 <p align="center">
   <img src="img/hello_world_tennis.png" width="500">
</p>

 It exists within an ecosystem of other great open source workflow management tools (like [Airflow](https://airflow.apache.org/), [Luigi](https://luigi.readthedocs.io/en/stable/), [Kubeflow](https://www.kubeflow.org/docs/about/kubeflow/) or [Prefect](https://docs.prefect.io/guide/)) while carving it's own niche based on the following design goals:

1. **Avoid unnecessary serialization:** `Primrose` keeps data in-memory between task steps, and only performs (de)serialization operations when explicitly requested by the user. Data is transported between nodes through use of a `DataObject` abstraction, which contextually delivers the correct data to each `Primrose` node at runtime. As a consequence of this design choice, `Primrose` runs on a single machine and can be deployed as a job within a single container, like any other Python script or cron job. In addition to operating on persistent data passed between nodes, `Primrose` can also be used to call external services in a manner similar to a [Luigi](https://luigi.readthedocs.io/en/stable/) job. In this way, Spark jobs or Hadoop scripts can be called and the framework simply dictates dependencies.

    * *As a comparison...* many solutions in this space are focused on long-running jobs which may be distributed across several computing nodes. Furthermore, to facilitate parallelization, save states for redundancy, and process datasets which are too large for memory, orchestrators often require that data is serialized between each workflow task. For smaller datasets, the IO time associated with these steps can be much longer than the time spent in computation.

    * *Primrose is not...* a solution which scales across clusters or a complex dependency management solution with dynamic DAGs (yet).

2. **Batch processesing for ML:** `Primrose` was built to facilitate frequent batches of model training or predictions that must read and write from/to multiple sources. Rather than requiring users to define their DAG structure in Python code, `Primrose` adopts a `configuration-as-code` approach. `Primrose` users create implementations of node objects once, then any DAG structural modifications or parameterization changes are processed through configuration json files. This way, deployment changes to DAG operations (such as modifying a DAG to serve model predictions instead of training) can be handled purely through configuration files. This avoids the need to build new Python scripts for production modifications. Furthermore, `Primrose` nodes are based on common machine learning tasks to make data scientist's lives easier. This cuts down on development time for building new models and maximizes code re-use among projects and teams. See the modeling examples in the source and documentation for more info!
    * *As a comparison...* in `Primrose`, users simply need to specify in their configuration file that they want common ML operations to act on the `DataObject`. These ML operations can certainly be implemented by users in Luigi or Airflow, but we found operations such as test-train splits or classifier cross-validation to be so common that they warranted nodes pre-dedicated to these operations.  [Prefect](https://docs.prefect.io/guide/) has made some great strides in this area, and we encourage users to check their solution out.

    * *Primrose is not...* an auto-ml tool or machine-learning toolkit which implements its own algorithms. Any Python machine learning library can be used with `Primrose`, simply by building model or pipeline nodes that implement the user's choice of library.

3. **Simplicity:**

    **Standardization of deployments:** `Primrose` is meant to help make deployment and model building as simple as possible. From a developer operations perspective, it requires no external scheduler or cluster to run deployments. `Primrose` code can simply be containerized with a `primrose` Python entrypoint, and deployed as a job on a k8s or any other container management service.

    **Standardization of development:** From a software engineering perspective, another advantage of `Primrose` stems form the standardization of model and recommender code. Modifying feature engineering pipelines or adding recommender features is simplified by writing additions to self-contained `Primrose` nodes and making additions to a configuration file.

    * *As a comparison...* `Primrose` can be leveraged as a piece of a larger ETL job (a `Primrose` job could be a job within an Airflow DAG), or run on it's own as a self-contained, single node ETL job. Some orchestration solutions (Airflow, for example) require running persistent clusters and services for managing jobs.

    * *Primrose is not...* able to manage its own job scheduling or timing. This is left to user using k8s job scheduling or manual cron job assignments on a virtual machine.


There are many solutions in this space, and we encourage users to explore other options that may be most appropriate for their workflows. We view `Primrose` as a simple solution for managing production ML jobs.


## Getting Started

`Primrose` has a couple of optional tools:
* a PostgreSQL database reader
* a plotting tool
* an R-dataset reader. E.g., to read in R's iris dataset see config/example_read_r.json

These require a few external dependencies, prior to its installation. If interested in their functionality, follow the appropriate instructions for your OS below. Otherwise, you can proceed with the basic package installation.

### Installation

You can install the latest `Primrose` release via pypi
```
pip install primrose
```
or you can clone the repository and install via `setup.py`.
```
git clone https://github.com/ww-tech/primrose.git
cd primrose
python setup.py install
```

To install the complete `Primrose` package (after dependencies have been installed):
```
    pip install primrose[postgres, plotting]
```

To install `Primrose` with just the PostgreSQL option:

```
pip install primrose[postgres]
```

To install `primrose` with just the plotting option:

```
pip install primrose[plotting]
```

To install `primrose` with just the R dataset reading option:

```
pip install primrose[R]
```
### External dependenices

**PostgreSQL**

#### MacOSX

We recommend using homebrew to manage OS level external packages. If you do not already have homebrew installed, please visit [their website](https://brew.sh/).

Instructions:
1. Use homebrew to install `postgresql` library.
   ```
   brew install postgresql
   ```

2. Use `pip` to install `psycopg2`
   ```
    pip install psycopg2
   ```

#### Debian/Ubuntu

Instructions:
1. Install the `libpq-dev` library
   ```
   apt-get install libpq-dev
   ```

**Plotting**

#### MacOSX

We recommend using homebrew to manage OS level external packages. If you do not already have homebrew installed, please visit [their website](https://brew.sh/).

Instructions:
1. Use homebrew to install `graphviz` library.
   ```
   brew install graphviz
   ```
2. If you are using a virtual environment such as `Anaconda` or `virtualenv`, you may need to specify a backend for `matplotlib`.
   ```
   mkdir -p ~/.matplotlib && touch ~/.matplotlib/matplotlibrc
   echo backend: TkAgg >> ~/.matplotlib/matplotlibrc
    ```

#### Debian/Ubuntu

Instructions:
1. Install `graphviz` library.
   ```
   apt-get install graphviz
   ```
2. If you are using a virtual environment such as `Anaconda` or `virtualenv`, you may need to specify a backend for `matplotlib`.
   ```
   mkdir -p ~/.config/matplotlib && touch ~/.config/matplotlib/matplotlibrc
   echo backend: Agg >> ~/.config/matplotlib/matplotlibrc
   ```

## Next
You are now ready to run your first `primrose` jobs: [Getting Started](README_GETTING_STARTED.md).

## License
Copyright 2019 WW International, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

CHANGELOG.md
# Changelog

## [v1.0.9](https://github.com/ww-tech/primrose/tree/v1.0.9) (2020-01-10)

[Full Changelog](https://github.com/ww-tech/primrose/compare/v1.0.8...v1.0.9)

**Merged pull requests:**

- update TRAVIS\_TAG to trigger deployments [\#59](https://github.com/ww-tech/primrose/pull/59) ([kzecchini](https://github.com/kzecchini))
- update bumpversion command [\#58](https://github.com/ww-tech/primrose/pull/58) ([kzecchini](https://github.com/kzecchini))
- update travis regex release branch [\#57](https://github.com/ww-tech/primrose/pull/57) ([kzecchini](https://github.com/kzecchini))
- Fix docs index page [\#56](https://github.com/ww-tech/primrose/pull/56) ([briangrahamww](https://github.com/briangrahamww))
- travis build docs [\#54](https://github.com/ww-tech/primrose/pull/54) ([briangrahamww](https://github.com/briangrahamww))
- Bump version fix [\#52](https://github.com/ww-tech/primrose/pull/52) ([kzecchini](https://github.com/kzecchini))
- add pip cache for travis [\#51](https://github.com/ww-tech/primrose/pull/51) ([briangrahamww](https://github.com/briangrahamww))
- Bump version fix [\#50](https://github.com/ww-tech/primrose/pull/50) ([kzecchini](https://github.com/kzecchini))
- updating docs for v1.0.8 [\#47](https://github.com/ww-tech/primrose/pull/47) ([kzecchini](https://github.com/kzecchini))
- Slack success/error  notification [\#40](https://github.com/ww-tech/primrose/pull/40) ([parul-l](https://github.com/parul-l))
- Reader datasets from R language [\#34](https://github.com/ww-tech/primrose/pull/34) ([weightwatchers-carlanderson](https://github.com/weightwatchers-carlanderson))
- Auto bump version [\#26](https://github.com/ww-tech/primrose/pull/26) ([kzecchini](https://github.com/kzecchini))

## [v1.0.8](https://github.com/ww-tech/primrose/tree/v1.0.8) (2019-12-18)

[Full Changelog](https://github.com/ww-tech/primrose/compare/86be32b40f29764fe98cfda4344df8dd49c6a531...v1.0.8)

**Implemented enhancements:**

- Switch to jinja templating for config file fragements [\#10](https://github.com/ww-tech/primrose/issues/10)
- Yaml config support [\#9](https://github.com/ww-tech/primrose/pull/9) ([kzecchini](https://github.com/kzecchini))

**Closed issues:**

- allow training fraction = 1 for TrainTestSplit pipeline [\#38](https://github.com/ww-tech/primrose/issues/38)
- allow transformers which change number of feature columns in SklearnPreprocessingTransformer [\#37](https://github.com/ww-tech/primrose/issues/37)
- reorder yaml configuration to match json [\#24](https://github.com/ww-tech/primrose/issues/24)
- Use YAML, not JSON [\#18](https://github.com/ww-tech/primrose/issues/18)
- primrose installation clarification [\#17](https://github.com/ww-tech/primrose/issues/17)
- create cache/ directory before running tests [\#5](https://github.com/ww-tech/primrose/issues/5)
- More docs on running primrose in your project [\#3](https://github.com/ww-tech/primrose/issues/3)

**Merged pull requests:**

- bumping version: 1.0.6 -\> 1.0.8 [\#41](https://github.com/ww-tech/primrose/pull/41) ([kzecchini](https://github.com/kzecchini))
- Train test split updates [\#39](https://github.com/ww-tech/primrose/pull/39) ([kzecchini](https://github.com/kzecchini))
- Updating yml config files to preserve order [\#30](https://github.com/ww-tech/primrose/pull/30) ([reka-ww](https://github.com/reka-ww))
- Primrose Installation Modification \(Issue \#17\) [\#27](https://github.com/ww-tech/primrose/pull/27) ([parul-l](https://github.com/parul-l))
- Jinja templating config file fragments [\#25](https://github.com/ww-tech/primrose/pull/25) ([briangrahamww](https://github.com/briangrahamww))
- Add class path config and dynamic import/registration of external nodes [\#23](https://github.com/ww-tech/primrose/pull/23) ([kzecchini](https://github.com/kzecchini))
- Adding primrose version command [\#22](https://github.com/ww-tech/primrose/pull/22) ([weightwatchers-carlanderson](https://github.com/weightwatchers-carlanderson))
- Add pickle support [\#21](https://github.com/ww-tech/primrose/pull/21) ([briangrahamww](https://github.com/briangrahamww))
- v1.0.7 documentation updates [\#14](https://github.com/ww-tech/primrose/pull/14) ([mskarlinski-ww](https://github.com/mskarlinski-ww))
- Fix a bug for the case when a transform occurs without the "to\_numeric" keyword [\#13](https://github.com/ww-tech/primrose/pull/13) ([mskarlinski-ww](https://github.com/mskarlinski-ww))
- Fix broken local test [\#11](https://github.com/ww-tech/primrose/pull/11) ([briangrahamww](https://github.com/briangrahamww))
- Update cli docs for --node\_module [\#7](https://github.com/ww-tech/primrose/pull/7) ([kzecchini](https://github.com/kzecchini))
- Add travis [\#6](https://github.com/ww-tech/primrose/pull/6) ([kzecchini](https://github.com/kzecchini))
- Sql jinja templating [\#4](https://github.com/ww-tech/primrose/pull/4) ([kzecchini](https://github.com/kzecchini))
- Cli doc fixes [\#2](https://github.com/ww-tech/primrose/pull/2) ([mskarlinski-ww](https://github.com/mskarlinski-ww))
- moved template file locations and updated setup.py [\#1](https://github.com/ww-tech/primrose/pull/1) ([mskarlinski-ww](https://github.com/mskarlinski-ww))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


primrose/__init__.py
import click
import logging
import os
import sys
import importlib
import pkg_resources
import shutil
from zipfile import ZipFile


logging.basicConfig(format='%(asctime)s %(levelname)s %(filename)s %(funcName)s: %(message)s', level=logging.INFO)


def replace_line(file_name, line_num, text):
    """Replace a single line of a file with text

    Args:
        file_name (str): name of file
        line_num (int): line to replace
        text (str): string to replace at line_num

    Returns:
        Nothing, the file is modified and rewritten

    """
    lines = open(file_name, 'r').readlines()
    lines[line_num] = text
    out = open(file_name, 'w')
    out.writelines(lines)
    out.close()


@click.group()
def cli():
    """This is the command line interface for primrose.
    The command most people need is `primrose run` to run a primrose job.
    Type primrose commandname --help for more detailed help on a command"""
    pass


@click.command()
@click.option('--config', required=True, help="Path to Config file")
def validate(config):
    """Validate a primrose config"""
    from primrose.configuration.configuration import Configuration
    configuration = Configuration(config_location=config)


@click.command()
@click.option('--config', required=True, help="Path to Config file")
@click.option('--dry_run', default=False, type=bool, help="Config file")
@click.option('--runner', default="local", type=click.Choice(["remote", "local"], case_sensitive=False), help="Run the job locally or on a remote machine")
@click.option('--path', default=None, type=str, help="Path to local files to include when running remotely")
def run(config, dry_run, runner, path):
    """Run a primrose job"""

    if runner == 'local':
        from primrose.configuration.configuration import Configuration
        from primrose.dag_runner import DagRunner

        configuration = Configuration(config_location=config)
        DagRunner(configuration).run(dry_run=dry_run)
    
    if runner == 'remote':
        import base64
        import requests
        import json

        host = os.environ.get('PRIMROSE_HOST')
        port = os.environ.get('PRIMROSE_PORT')

        filetype = os.path.splitext(config)[-1]
        
        with open(config, 'rb') as f:
            config_str = f.read()
            b64_config = base64.encodebytes(config_str).decode()
        
        if path is not None:
            try:
                # create a ZipFile object
                zip_file_name = "tmp.zip"
                url = f'{host}:{port}/run-custom-dag'

                with ZipFile(zip_file_name, 'w') as zip_obj:
                    for folder_name, _, filenames in os.walk(path):
                        for filename in filenames:
                            file_path = os.path.join(folder_name, filename)
                            zip_obj.write(file_path)
                        
                zip_obj = open(zip_file_name, 'rb')
                r = requests.post(url, data={"b64_config": b64_config, "dry_run": dry_run, "filetype": filetype}, files={"in_file": zip_obj})

            except Exception as e:
                raise e
            finally:
                os.remove(zip_file_name)

        else:
            url = f'{host}:{port}/run-base-dag'
            print({"b64_config": b64_config, "dry_run": dry_run, "filetype": filetype})
            r = requests.post(url, data=json.dumps({"b64_config": b64_config, "dry_run": dry_run, "filetype": filetype}))
        
        try:
            job_id = r.json().get("job_id")
            logging.info(f'primrose job_id: {job_id}')
            logging.info(f'')
            logging.info(f'Streaming logs from remote job')
            logging.info(f'Press CTRL-C to stop log stream')
            logging.info(f'Exiting log stream will not cancel the job')
            logging.info(f'')
            logs_r = requests.get(f'{host}:{port}/log-stream/{job_id}', stream=True)
            for line in logs_r.iter_content(4096):
                if line:
                    decoded_line = line.decode('utf-8')
                    logging.info(decoded_line)
        except KeyboardInterrupt:
            logging.info(f'')
            logging.info(f'Log stream interrupted - job will continue to process')
            logging.info(f'')


@click.command()
@click.option('--config', required=True, help="Path to Config file")
@click.option('--node_size', help='Size of nodes', required=False, default=500)
@click.option('--label_font_size', help='Size of font for text labels', required=False, default=12)
@click.option('--text_angle', help='Angle to rotate text (counterclockwise)', required=False, default=0)
@click.option('--image_width', help='width of image in inches', required=False, default=16)
@click.option('--image_height', help='height of image in inches', required=False, default=12)
@click.option('--nodesequence', help='Show node sequences numbers? "true" or "false"?', required=False)
@click.option('--outfile', help='Path for the output image file', required=True)
def plot(config, node_size, label_font_size, text_angle, image_width, image_height, nodesequence, outfile):
    '''Create an image of the DAG'''
    from primrose.configuration.configuration import Configuration
    from primrose.dag.traverser_factory import TraverserFactory

    config = Configuration(config_location=config)

    filename = outfile
    other_args = {}
    other_args['filename'] = outfile

    if node_size:
        other_args['node_size'] = int(node_size)

    if label_font_size:
        other_args['label_font_size'] = int(label_font_size)

    if text_angle:
        other_args['text_angle'] = int(text_angle)

    if image_width:
        other_args['image_width'] = int(image_width)

    if image_height:
        other_args['image_height'] = int(image_height)

    other_args['traverser'] = None
    if nodesequence and nodesequence.lower() == "true":
        traverser = TraverserFactory().default_traverser(config)

        if config.config_metadata and 'traverser' in config.config_metadata:
            traverser = TraverserFactory().instantiate(config.config_metadata['traverser'], config)
            logging.info("Setting to Traverser to %s", config.config_metadata['traverser'])

        other_args['traverser'] = traverser

    logging.info("passing in %s", other_args)
    config.dag.plot_dag(**other_args)


@click.command()
@click.option('--destination', required=True, help="Path to destination")
def generate_run_script(destination):
    """Create primrose run script in your project--if you need to register your own primrose classes.
    This will also allow you to register and use your own primrose node classes. 
    To register your own classes you will also need to run the primrose generate_class_registration_template command"""
    from shutil import copyfile
    run_primrose_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'templates', 'run_primrose.py')
    copyfile(run_primrose_path, destination)
    print("Primrose's run_primrose.py script copied to " + destination)
    print("Don't forget to set where your new classes are to be imported. See script.")


@click.command()
@click.option('--destination', required=True, help="Path to destination")
def generate_class_registration_template(destination):
    """Create template to register your own classes. This will copy a template to your destination, such as src/__init__.py"""
    from shutil import copyfile
    user_registration_template_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
                                                   'templates',
                                                   'user_registration_template.py')
    copyfile(user_registration_template_path, destination)
    print("Primrose's user_registration_template.py script copied to " + destination)

@click.command()
@click.option('--name', required=True, help="Name of primrose project to be generated")
def create_project(name):
    """Create template project filled with example configuration"""

    from shutil import copyfile, copytree

    # get local directories to copy from the package dir
    package_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'templates')
    env_dir = sys.prefix

    config_dir_path = os.path.join(env_dir, 'config')

    data_dir_path = os.path.join(env_dir, 'data')

    user_registration_template_path = os.path.join(package_dir, 'user_registration_template.py')

    run_primrose_path = os.path.join(package_dir, 'run_primrose.py')

    # make empty cache and src dir
    os.makedirs(os.path.join(name, 'cache'))
    os.makedirs(os.path.join(name, 'src', 'yourpackage'))

    # copy relevant files to user path
    copytree(config_dir_path, os.path.join(name, 'config'))

    copytree(data_dir_path, os.path.join(name, 'data'))

    copyfile(os.path.join(package_dir, 'awesome_model.py'), os.path.join(name,
                                                                         'src',
                                                                         'yourpackage',
                                                                         'awesome_model.py'))

    copyfile(os.path.join(package_dir, 'awesome_reader.py'), os.path.join(name,
                                                                          'src',
                                                                          'yourpackage',
                                                                          'awesome_reader.py'))

    copyfile(user_registration_template_path, os.path.join(name, 'src', '__init__.py'))

    copyfile(run_primrose_path, os.path.join(name, 'run_primrose.py'))

    # modify run_primrose to take into account the templated extention functions
    replacement_line = 'from src.__init__ import *\n'

    replace_line(os.path.join(name, 'run_primrose.py'), 14, replacement_line)

    # make __init__ file for the yourpackage directory
    with open(os.path.join(name, 'src', 'yourpackage','__init__.py'), 'w') as f:
        f.write('')

    print("New primrose project, {} built!".format(name))

@cli.command()
def version():
    '''Print the installed primrose version'''
    print(pkg_resources.get_distribution("primrose").version)

cli.add_command(validate)
cli.add_command(run)
cli.add_command(plot)
cli.add_command(generate_run_script, name='generate-run-script')
cli.add_command(generate_class_registration_template, name='generate-class-registration-template')
cli.add_command(create_project, name='create-project')
cli.add_command(version)

if __name__ == "__main__":
    cli()
